File: NX1926_NX1930_NX1934_NX1938_NX1942_NX1946/UGOPEN/SampleNXOpenApplications/Java/Routing/Routing_MaximumHeight_DesignRule.java

/*==================================================================================================
               Copyright 2017 Siemens Product Lifecycle Management Software Inc.
====================================================================================================

    NOTE:  NX Development provides programming examples for illustration only.
    NX Development assumes you are familiar with the programming language
    demonstrated in these examples, and the tools used to create and debug NX/Open
    programs. GTAC support professionals can help explain the functionality of
    a particular procedure, but neither GTAC nor NX Development will modify
    these examples to provide added functionality or construction procedures.

File description:

    A Java class that implements Maximum Height Violation Routing Design Rule.


==================================================================================================*/
import java.rmi.RemoteException;
import java.util.List;
import java.util.Vector;

import nxopen.ListingWindow;
import nxopen.LogFile;
import nxopen.NXException;
import nxopen.NXObject;
import nxopen.Session;
import nxopen.SessionFactory;
import nxopen.routing.CustomManager;
import nxopen.routing.ISegment;
import nxopen.routing.Overstock;
import nxopen.routing.OverstockApplication;
import nxopen.routing.OverstockApplicationCollection;
import nxopen.routing.RouteManager;
import nxopen.routing.StockDefinition;
import nxopen.routing.Wire;
import nxopen.routing.electrical.*;
import nxopen.routing.StockDevice; 
import nxopen.routing.CustomManager.DesignRuleReason;

class MaxHeightDesignRules implements nxopen.routing.CustomManager.DesignRule
{
    static MaxHeightDesignRules theMaxHeightDesignRules = null;
    static Session theSession = null;
    static ListingWindow listingWindow = null;
    
    //------------------------------------------------------------------------------------------
    // Called when NX starts.
    // Creates a new instance of this Callback class.
    public static void startup( String [] args ) throws NXException, java.rmi.RemoteException
    {     
     
        if (theSession == null)
            theSession = (Session)SessionFactory.get((String)"Session");

        if (listingWindow == null)
            listingWindow = theSession.listingWindow();
        
        theMaxHeightDesignRules = new MaxHeightDesignRules();
    
    }
    
    //------------------------------------------------------------------------------------------
    // The constructor for this class.
    // Registers the Design Rules.
    public MaxHeightDesignRules() throws RemoteException, NXException
    {        
        try
        {
            CustomManager customManager = theSession.routeCustomManager();  
            String shortDescription = "Maximum Height Violation";
            String longDescription = "The Maximum Height design rule makes sure the diameter of each individual connection is within the permissible limit of value specified by attribute CABLETRAY_MAXHEIGHT."; 
            customManager.addDesignRule(DesignRuleReason.CREATE_PATH,shortDescription,longDescription , this);
            customManager.addDesignRule(DesignRuleReason.HEAL_PATH,shortDescription,longDescription , this);
            customManager.addDesignRule(DesignRuleReason.ASSIGN_CORNER,shortDescription,longDescription , this);
            customManager.addDesignRule(DesignRuleReason.REMOVE_PART,shortDescription,longDescription , this);
            customManager.addDesignRule(DesignRuleReason.AUTO_ROUTE_PIN_LEVEL,shortDescription,longDescription , this);
            customManager.addDesignRule(DesignRuleReason.MANUAL_ROUTE_PIN_LEVEL,shortDescription,longDescription , this);
            customManager.addDesignRule(DesignRuleReason.AUTO_ROUTE_COMPONENT_LEVEL,shortDescription,longDescription , this);
            customManager.addDesignRule(DesignRuleReason.MANUAL_ROUTE_COMPONENT_LEVEL,shortDescription,longDescription , this);
            customManager.addDesignRule(DesignRuleReason.INTERACTIVE,shortDescription,longDescription , this);
            customManager.addDesignRule(DesignRuleReason.TRANSFORM_PATH,shortDescription,longDescription , this);            
            customManager.addDesignRule(DesignRuleReason.EDIT_CHARACTERISTIC,shortDescription,longDescription , this);
        }
        catch ( NXException ex)
        {
            theSession.logFile().writeLine( "\n\n\n***Error in Routing_MaximumHeight_DR constructor: " + ex.getMessage() );
        }
        catch (RemoteException ex)
        {
            theSession.logFile().writeLine( "\n\n\n***RemoteException in Routing_MaximumHeight_DR constructor: " + ex.getMessage() );
        }        
    }
    
    //---------------------------------------------------------------------------------------------
    // This is the Create Path Design Rule.
    //
    // It simply prints the tags of the given object to the Listing Window and Log File.
    public void designRule(DesignRuleReason reason, NXObject[] objects) throws NXException, RemoteException 
    {
        CustomManager customManager = theSession.routeCustomManager();
        
        // For interactive checks, check all the segments in the work part.
        if (reason == CustomManager.DesignRuleReason.INTERACTIVE)
        {
            nxopen.TaggedObjectCollection.Iterator objectCollectionItr = theSession.parts().work().segmentManager().segments().iterator();

             while(objectCollectionItr.hasNext())
             {
              ISegment segment = (ISegment) objectCollectionItr.next();
              checkForMaximumHeightViolation(theSession.parts().work().routeManager(),customManager,segment, reason);
             }
             return;
        }
        
        // Otherwise, check the segments in the given list of objects.
        for (NXObject nxObject : objects)
        {                
            if(nxObject instanceof ISegment)
            {
                ISegment segment = (ISegment)nxObject;
                checkForMaximumHeightViolation(theSession.parts().work().routeManager(), customManager, segment, reason);
            }
        }
    }
    //--------------------------------------------------------------------------------------------------
    // Get the the double value from attribute
    //
    // \param[in]
    //      attributeInfo
    //         Attribute information from which double needs to be queried
    //
    // \return
    //      Value of attribute as Double
    private double convertToReal
    (
        NXObject.AttributeInformation attributeInfo
    )throws RemoteException, NXException
    {
        switch(attributeInfo.type.ordinal())
        {
            case    NXObject.AttributeType._REAL:
                    return attributeInfo.realValue;
                    
            case    NXObject.AttributeType._INTEGER:
                return (double)attributeInfo.integerValue;
                
            case    NXObject.AttributeType._STRING:
                return Double.parseDouble(attributeInfo.stringValue);
                
            default:
                NXException.create("Unknown attribute type");            
        }
        return 0;
    }
    //--------------------------------------------------------------------------------------------------
    // Gets the real value of the attribute on the given object, if any.
    //
    // \param[in]
    //      object
    //          The object to check for the attribute.
    //
    // \param[in]
    //      attributeName
    //          The name (a.k.a. title) of the attribute for which to look.
    //
    // \return
    //      Value of attribute as Double
    //
    private double getAttributeValue
    (
        ISegment              segment,
        String                attributeNameString
    ) throws RemoteException, NXException
    {        
        NXObject.AttributeInformation[] attributeInformationArray = ((NXObject) segment).getUserAttributes();
        
        for(NXObject.AttributeInformation attributeInfo : attributeInformationArray)
        {
            if(attributeInfo.title.equals(attributeNameString))
            {                
                return convertToReal(attributeInfo);
            }
        }
        return 0;
    }
    //--------------------------------------------------------------------------------------------------
    // Queries the list of stock devices routing through the segment
    //
    // \param[in]
    //      segment
    //          Segment through which stock devices are routed.
    //
    // \param[in]
    //      routeManager
    //          RouteManager from the current work part
    //
    // \return
    //          List of stock stock devices referencing the segment
    //
    private List<StockDevice> getStockDevicesRoutedThroughSegment
    (
        ISegment              segment,
        RouteManager          routeManager
    ) throws RemoteException, NXException
    {        
        List<StockDevice> stockDeviceList = new Vector<StockDevice>();
        ElectricalStockDeviceCollection stockDevices = routeManager.electricalStockDevice();

        nxopen.TaggedObjectCollection.Iterator stockDevicesItr = stockDevices.iterator();
        
        while(stockDevicesItr.hasNext())
        {
            StockDevice stockDevice = (StockDevice) stockDevicesItr.next();
            
            Wire wire = null;
            wire = (Wire)stockDevice.nxEquivalent();
            
            if (wire == null) continue;
            
            ISegment[] stockDeviceSegments = wire.getSegmentList();

            for(ISegment singleSegment : stockDeviceSegments)
            {                
                if (! singleSegment.equals(segment))
                {
                    continue;
                }
                
                if(!stockDeviceList.contains(stockDevice))
                {
                    stockDeviceList.add(stockDevice);
                }
            }
        }
        
        return stockDeviceList;
    }    
    //-----------------------------------------------------------------------------------------------
    // This method gives diameter of maximum size connection amongst the stock devices routing through the segment
    //
    // \param[in]
    //      segment
    //          Segment through which connections are routed
    // \param[in]
    //      stockDevicesOnSegment
    //          List of stock devices routed through the segment
    // \param[out]
    //      largestDiameter
    //          Maximum size diameter 
    private double getValueToCheckMaxHeightViolationForStockDevice
    (
        ISegment                      segment,
        List <StockDevice>            stockDevicesOnSegment
    ) throws RemoteException, NXException
    {
        double valueToCheck = 0;
        
        for (StockDevice stockDevice :stockDevicesOnSegment)
        {
            double maxValue = 0.0;
            
            StockDefinition stockDef = null;
            ElectricalStockDevice electricalStockDevice = null;
            
            if(stockDevice instanceof ElectricalStockDevice)
            {
                electricalStockDevice = (ElectricalStockDevice)stockDevice;
            }

            if(electricalStockDevice == null)
            {
                continue;
            }

            CableDefinition cableDefinition = electricalStockDevice.findTopmostCableDefinition();
            if (cableDefinition != null)
            {             
                CableDevice cableDevice = null;
                
                if(cableDefinition.getReferencingDevice() instanceof CableDevice)
                {
                    cableDevice = (CableDevice)cableDefinition.getReferencingDevice();
                }
                
                if (cableDevice == null)
                {
                    continue;
                }
                stockDef = cableDevice.stockDefinition();
            }
            else
            {
                stockDef = stockDevice.stockDefinition();
            }

            maxValue = stockDef.outerDiameter();
            
            if (Double.compare(maxValue, valueToCheck) > 0)
            {
                valueToCheck = maxValue;
            }
        }        
        return valueToCheck;        
    }    
    //-----------------------------------------------------------------------------------------------
    // Queries the overstock referencing the segment
    //
    // \param[in]
    //      segment
    //         Segment from which referencing overstock needs to be queried.
    // \param[in]
    //      routeManager
    //          RouteManager from the current work part
    // \return
    //          Set of overstocks referencing the segment
    private List<Overstock> getOverstocksPresentOnTheSegment
    (
        ISegment                     segment,
        RouteManager                 routeManager
    ) throws RemoteException, NXException
    {
         List<Overstock> overstockList = new Vector<Overstock>();
         
         OverstockApplicationCollection overstockApplications = routeManager.overstockApplications();
         nxopen.TaggedObjectCollection.Iterator overstockApplicationsItr = overstockApplications.iterator();
         
         while(overstockApplicationsItr.hasNext())
         {
             OverstockApplication overstockApplication = (OverstockApplication) overstockApplicationsItr.next();
             ISegment[] overstockApplicationSegments = overstockApplication.getSegments();

             for (ISegment currentSegment : overstockApplicationSegments)
             {
                 if (currentSegment != segment)
                 {
                     continue;
                 }

                 Overstock[] overstocks = overstockApplication.getOverstocks();
                 overstockList.add(overstocks[0]);
             }
         }
         return overstockList;
    }
    //------------------------------------------------------------------------------------------
    // Check the maximum diameter of connection routing through given segment against the maximum allowed height.
    //
    // Creates a violation for the given reason if connection diameter exceeds specified Maximum Height value.
    //
    // \param[in]
    //      routeManager
    //          RouteManager from current work part
    //
    // \param[in]
    //      customManager
    //          CustomManager to create design Rule violation.
    //
    // \param[in]
    //      segment
    //          Routing Segment to be check for Maximum Height violation
    //
    // \param[in]
    //      reason
    //          Design Rule reason for which Maximum Height Violation needs to be checked.
    private void checkForMaximumHeightViolation
    ( 
        RouteManager                    routeManager, 
        CustomManager                   customManager, 
        ISegment                        segment, 
        CustomManager.DesignRuleReason  reason
    ) throws RemoteException, NXException
    {
        // Get value of allowed Maximum Height or Specified Maximum Height
        String maxHeightAttributeName = "CABLETRAY_MAXHEIGHT";
        double specifiedMaxHeightValue = getAttributeValue(segment,maxHeightAttributeName);     
        if(Math.abs(specifiedMaxHeightValue) < 1e-5)
        {
         return;
        }    
     
        // Get list of Stock Devices routing through the segment
        List<StockDevice> stockDeviceList =  getStockDevicesRoutedThroughSegment(segment,routeManager);
       
        if(stockDeviceList.isEmpty())
        {
             return;
        }
        
        // Get max value diameter of connection routing through the segment
        double valueToCompare = getValueToCheckMaxHeightViolationForStockDevice(segment,stockDeviceList);
        
        // Get Overstocks referencing the segment
        List<Overstock> overstockList = getOverstocksPresentOnTheSegment(segment,routeManager);
        
        
        // If Overstocks are present on segment, get maximum value of diameter including overstock
        if(overstockList.size() > 0)
        {
            double totalOverstockThickness = 0.0;
            for(Overstock overstock : overstockList)
            {
                totalOverstockThickness += overstock.getThickness().value();
            }
            valueToCompare += (2 * totalOverstockThickness);
        }
        
        // If current value of diameter is greater than specified maximum height value then raise violation.
        if (Double.compare(valueToCompare, specifiedMaxHeightValue) > 0)
        { 
            String description = "The stock exceeds the maximum height allowed for this segment. The current height is " + valueToCompare + ".The allowed height is " + specifiedMaxHeightValue + ".";

            NXObject[] objects = {(NXObject)segment };
            customManager.createViolationForReason(reason,"The segment violates Maximum Height Rule", description, objects);
        }

        stockDeviceList.clear();
        overstockList.clear();     
    }
    //------------------------------------------------------------------------------------------
    // Tells NX when to unload your application.
    // For Routing callbacks, plugins, and Design Rules, this MUST return AtTermination.
    public static int getUnloadOption(String arg)
    {
        return Session.LibraryUnloadOption.AT_TERMINATION;
    }
}

File: NX2007/UGOPEN/SampleNXOpenApplications/Python/Routing/Routing_MaximumHeight_DesignRule.py
#===================================================================================================
#               Copyright 2017 Siemens Product Lifecycle Management Software Inc.
#===================================================================================================
#
#    NOTE:  NX Development provides programming examples for illustration only.
#    NX Development assumes you are familiar with the programming language
#    demonstrated in these examples, and the tools used to create and debug NX/Open
#    programs. GTAC support professionals can help explain the functionality of
#    a particular procedure, but neither GTAC nor NX Development will modify
#    these examples to provide added functionality or construction procedures.
#
#File description:
#
#    A python script that implements Maximum Height Violation Routing Design Rule.
#
#
#===================================================================================================
import NXOpen
import NXOpen.Routing

#---------------------------------------------------------------------------------------------------
# Get the the double value from attribute
#
# \param[in]
#      attributeInfo
#         Attribute information from which double needs to be queried
#
# \return
#      Value of attribute as Double
#
def convertToReal( attribute ):

    if attribute.Type == NXOpen.NXObject.AttributeType.Real:
        return attribute.RealValue

    if attribute.Type == NXOpen.NXObject.AttributeType.Integer:
        return float( attribute.IntegerValue )

    if attribute.Type == NXOpen.NXObject.AttributeType.String:
        return float( attribute.StringValue )

    raise ValueError( 'Unexpected attribute type.' )

#---------------------------------------------------------------------------------------------------
# Gets the real value of the attribute on the given object, if any.
#
# \param[in]
#      object
#          The object to check for the attribute.
#
# \param[in]
#      attributeName
#          The name (a.k.a. title) of the attribute for which to look.
#
# \return
#          The value of the attribute, if found.
def getAttributeValue( nxObjectTag, attributeNameString ):

    nxObject = NXOpen.TaggedObjectManager.GetTaggedObject( nxObjectTag )

    attrInfoVector = nxObject.GetUserAttributes()

    for attribute in attrInfoVector:

        if attribute == None:
            continue

        if attribute.Title == attributeNameString:
            return convertToReal( attribute )

    return 0.0

#---------------------------------------------------------------------------------------------------
# Queries the list of stock devices routing through the segment
#
# \param[in]
#      segment
#          Segment through which stock devices are routed.
#
# \param[in]
#      routeManager
#          RouteManager from the current work part
#
# \return
#          List of stock stock devices referencing the segment
#
def getStockDevicesRoutedThroughSegment( segmentTag, routeManager ):
    
    stockDevicesOnSegment = []

    stockDeviceCollection = routeManager.ElectricalStockDevice

    for stockDevice in stockDeviceCollection:
        wire = stockDevice.NxEquivalent
        if wire == None:
            continue

        for singleSegment in wire.GetSegments():
            if singleSegment.Tag == segmentTag:
                stockDevicesOnSegment.append( stockDevice )

    # Use a temporary set to remove duplicate stock devices, if any.
    tempSet = set( stockDevicesOnSegment )
    stockDevicesOnSegment = list( tempSet )

    return stockDevicesOnSegment
#------------------------------------------------------------------------------------------------
# This method gives diameter of maximum size connection amongst the stock devices routing through the segment
#
# \param[in]
#      segment
#          Segment through which connections are routed
# \param[in]
#      stockDevicesOnSegment
#          List of stock devices routed through the segment
# \return
#          Maximum size diameter 
def getValueToCheckMaxHeightViolationForStockDevice( segmentTag, stockDevicesOnSegment ):

    valueToCompare = 0.0

    for stockDevice in stockDevicesOnSegment:
        maxValue = 0.0

        cableDefinition = stockDevice.FindTopmostCableDefinition()

        stockDef = None

        if cableDefinition == None:
            stockDef = stockDevice.StockDefinition
        else:
            cableDevice = cableDefinition.GetReferencingDevice()
            stockDef    = cableDevice.StockDefinition

        maxValue = stockDef.OuterDiameter
        
        if maxValue > valueToCompare:
            valueToCompare = maxValue

    return valueToCompare

#------------------------------------------------------------------------------------------------
# Queries the overstock referencing the segment
#
# \param[in]
#      segment
#         Segment from which referencing overstock needs to be queried.
# \param[in]
#      routeManager
#          RouteManager from the current work part
# \return
#          Set of overstocks referencing the segment
def getOverstocksPresentOnTheSegment( segmentTag, routeManager ):
    
    overstockOnSegment = []

    for overstockApplication in routeManager.OverstockApplications:
        
        segmentsOfOverstockApplication = overstockApplication.GetSegments()

        for currentSegment in segmentsOfOverstockApplication:

            if singleSegment.Tag == segmentTag:
                overstocks = overstockApplication.GetOverstocks()
                overstockOnSegment.append( overstocks[0] )

    return overstockOnSegment;

#---------------------------------------------------------------------------------------------------
# Check the maximum diameter of connection routing through given segment against the maximum allowed height.
#
# Creates a violation for the given reason if connection diameter exceeds specified Maximum Height value.
#
# \param[in]
#      routeManager
#          RouteManager from current work part
#
# \param[in]
#      customManager
#          CustomManager to create design Rule violation.
#
# \param[in]
#      segment
#          Routing Segment to be check for Maximum Height violation
#
# \param[in]
#      reason
#          Design Rule reason for which Maximum Height Violation needs to be checked.
def checkForMaximumHeightViolation( routeManager, customManager, segment, reason ):
    
    maxHeightAttributeName = "CABLETRAY_MAXHEIGHT"
    
    # Get value of allowed Maximum Height or Specified Maximum Height
    specifiedMaxHeightValue = getAttributeValue( segment.Tag, "CABLETRAY_MAXHEIGHT" )
    if specifiedMaxHeightValue <= 0.0:
        return

    stockDevices = getStockDevicesRoutedThroughSegment( segment.Tag, routeManager )
    if len( stockDevices ) == 0:
        return

    # Get max value diameter of connection routing through the segment
    valueToCompare = getValueToCheckMaxHeightViolationForStockDevice( segment.Tag, stockDevices )            
    
    # Get Overstocks referencing the segment
    overstocksReferencingTheSegment = getOverstocksPresentOnTheSegment( segment.Tag, routeManager )
    
    # If Overstocks are present on segment, get maximum value of diameter including overstock
    if len( overstocksReferencingTheSegment ) > 0:
    
        totalOverstockThickness = 0.0;
        for overstock in overstocksReferencingTheSegment:
            totalOverstockThickness += overstock.GetThickness().Value
        
        valueToCompare += (2 * totalOverstockThickness)

    # If current value of diameter is greater than specified maximum height value then raise violation.
    if valueToCompare > specifiedMaxHeightValue:
    
        description = "The stock exceeds the maximum height allowed for this segment. The current height is " + str( valueToCompare ) + ".The allowed height is " + str( specifiedMaxHeightValue ) + "."

        objects = [segment]
        customManager.CreateViolationForReason( "Maximum Height Violation", reason, "The segment violates Maximum Height Rule", description, objects);

#---------------------------------------------------------------------------------------------------
def createMaximumHeightDesignRule( reason, objects ):
    try:

        session        = NXOpen.Session.GetSession()
        routeManager   = session.Parts.Work.RouteManager
        customManager  = session.RouteCustomManager

        # For interactive checks, check all the segments in the work part.
        if reason == NXOpen.Routing.CustomManager.DesignRuleReason.Interactive:
            
            segmentCollection = session.Parts.Work.SegmentManager.Segments

            for segment in segmentCollection:
                checkForMaximumHeightViolation( routeManager, customManager, segment, reason )

            return
        
        # Otherwise, check the segments in the given list of objects.
        for object in objects:
            if type( object ) is NXOpen.Routing.ArcSegment or \
               type( object ) is NXOpen.Routing.LineSegment or \
               type( object ) is NXOpen.Routing.SplineSegment:
                checkForMaximumHeightViolation( routeManager, customManager, object, reason )

    except Exception as ex:
        NXOpen.UI.GetUI().NXMessageBox.Show( "Python", NXOpen.NXMessageBox.DialogType.Error, str( ex ) )
        raise ex

#---------------------------------------------------------------------------------------------------
def registerMaxHeightDesignRules():

    maxHeightViolationTitle       = "Maximum Height Violation"
    maxHeightViolationDescription = "The Maximum Height design rule makes sure the diameter of each individual connection is within the permissible limit of value specified by attribute CABLETRAY_MAXHEIGHT."
        
    customManager = NXOpen.Session.GetSession().RouteCustomManager
        
    customManager.AddDesignRule(NXOpen.Routing.CustomManager.DesignRuleReason.CreatePath,                maxHeightViolationTitle, maxHeightViolationDescription, createMaximumHeightDesignRule )
    customManager.AddDesignRule(NXOpen.Routing.CustomManager.DesignRuleReason.HealPath,                  maxHeightViolationTitle, maxHeightViolationDescription, createMaximumHeightDesignRule )
    customManager.AddDesignRule(NXOpen.Routing.CustomManager.DesignRuleReason.AssignCorner,              maxHeightViolationTitle, maxHeightViolationDescription, createMaximumHeightDesignRule )
    customManager.AddDesignRule(NXOpen.Routing.CustomManager.DesignRuleReason.AssignStock,               maxHeightViolationTitle, maxHeightViolationDescription, createMaximumHeightDesignRule )
    customManager.AddDesignRule(NXOpen.Routing.CustomManager.DesignRuleReason.RemovePart,                maxHeightViolationTitle, maxHeightViolationDescription, createMaximumHeightDesignRule )
    customManager.AddDesignRule(NXOpen.Routing.CustomManager.DesignRuleReason.AutoRoutePinLevel,         maxHeightViolationTitle, maxHeightViolationDescription, createMaximumHeightDesignRule )
    customManager.AddDesignRule(NXOpen.Routing.CustomManager.DesignRuleReason.ManualRoutePinLevel,       maxHeightViolationTitle, maxHeightViolationDescription, createMaximumHeightDesignRule )
    customManager.AddDesignRule(NXOpen.Routing.CustomManager.DesignRuleReason.AutoRouteComponentLevel,   maxHeightViolationTitle, maxHeightViolationDescription, createMaximumHeightDesignRule )
    customManager.AddDesignRule(NXOpen.Routing.CustomManager.DesignRuleReason.ManualRouteComponentLevel, maxHeightViolationTitle, maxHeightViolationDescription, createMaximumHeightDesignRule )
    customManager.AddDesignRule(NXOpen.Routing.CustomManager.DesignRuleReason.TransformPath,             maxHeightViolationTitle, maxHeightViolationDescription, createMaximumHeightDesignRule )
    customManager.AddDesignRule(NXOpen.Routing.CustomManager.DesignRuleReason.EditCharacteristic,        maxHeightViolationTitle, maxHeightViolationDescription, createMaximumHeightDesignRule )

#---------------------------------------------------------------------------------------------------
def startup( argc, args ):
    try:

        registerMaxHeightDesignRules()

    except Exception as ex:
        NXOpen.UI.GetUI().NXMessageBox.Show( "Python", NXOpen.NXMessageBox.DialogType.Error, str( ex ) )
        raise ex

